### PR TITLE
Steve - L104_bcoc_en_USA

### PR DESCRIPTION
**Pull request by ssmithClimate**
#282

### CODE REVIEW

File: R/zchunk_L104.bcoc_en_USA_S_T_Y.R [hunk 1]
```diff
@@ -1,25 +1,25 @@
...
 #' the generated outputs: \code{L104.bcoc_tgej_USA_en_T_1990}. The corresponding file in the
 #' original data system was \code{L104.bcoc_en_USA_S_T_Y.R} (emissions level1).
-#' @details Describe in detail what this chunk does.
+#' @details 1990 USA BC and OC emissions are divided by year 2000 USA energy use to generate emission factors for GCAM aggregate sectors
 #' @importFrom assertthat assert_that
 #' @importFrom dplyr filter mutate select
 #' @importFrom tidyr gather spread
-#' @author YourInitials CurrentMonthName 2017
+#' @author SJS May 2017
 #' @export
```
Comment (bpbond) [inline R/zchunk_L104.bcoc_en_USA_S_T_Y.R]: Remove `@export` line when all done

File: R/zchunk_L104.bcoc_en_USA_S_T_Y.R [hunk 2]
```diff
@@ -32,49 +32,130 @@ module_emissions_L104.bcoc_en_USA_S_T_Y_DISABLED <- function(command, ...) {
...
-    #
-    # if(OLD_DATA_SYSTEM_BEHAVIOR) {
-    #   ... code that replicates old, incorrect behavior
-    # } else {
-    #   ... new code with a fix
-    # }
+	# Select for USA data
+	Bond_BC_1990 %>%
+    select(Technology, USA) %>%
+	  rename(BC_em = USA) -> Bond_BC_USA_1990
```
Comment (bpbond) [inline R/zchunk_L104.bcoc_en_USA_S_T_Y.R]: I think you can one-line this if you want: `select(Technology, BC_em = USA)`

File: R/zchunk_L104.bcoc_en_USA_S_T_Y.R [hunk 3]
```diff
@@ -32,49 +32,130 @@ module_emissions_L104.bcoc_en_USA_S_T_Y_DISABLED <- function(command, ...) {
...
+
+	Bond_OC_1990 %>%
+    select(Technology, USA) %>%
+	  rename(OC_em = USA)  -> Bond_OC_USA_1990
+
+	# Combine BC & OC data and convert to Tg
+	Bond_OC_USA_1990 %>%
+	  left_join_error_no_match(Bond_BC_USA_1990, by=("Technology")) %>%
```
Comment (bpbond) [inline R/zchunk_L104.bcoc_en_USA_S_T_Y.R]: Spaces around `=` and other math operators pls

File: R/zchunk_L104.bcoc_en_USA_S_T_Y.R [hunk 4]
```diff
@@ -32,49 +32,130 @@ module_emissions_L104.bcoc_en_USA_S_T_Y_DISABLED <- function(command, ...) {
...
+	  mutate(OC_em = OC_em * CONV_GG_TG)  %>%
+	  mutate(BC_em = BC_em * CONV_GG_TG) ->
```
Comment (bpbond) [inline R/zchunk_L104.bcoc_en_USA_S_T_Y.R]: `mutate(OC_em = ..., BC_em = ...)`

File: R/zchunk_L104.bcoc_en_USA_S_T_Y.R [hunk 5]
```diff
@@ -32,49 +32,130 @@ module_emissions_L104.bcoc_en_USA_S_T_Y_DISABLED <- function(command, ...) {
...
+	  Bond_BCOC_USA_1990
+
+	# Aggregate by GCAM sector and technology
+	Bond_BCOC_USA_1990 %>%
+	  rename(BCOC.Technology = Technology) %>%
+	  # Not all items in the BCOC input dataset have GCAM equivalents, so will have some NAs. 
+  	  # This is expected. Use left_join for this reason.
```
Comment (bpbond) [inline R/zchunk_L104.bcoc_en_USA_S_T_Y.R]: Nice commenting! Thank you!

File: R/zchunk_L104.bcoc_en_USA_S_T_Y.R [hunk 6]
```diff
@@ -32,49 +32,130 @@ module_emissions_L104.bcoc_en_USA_S_T_Y_DISABLED <- function(command, ...) {
...
+	  left_join(Bond_bc.oc_tech,by = c("BCOC.Technology")) %>%
+	  group_by(sector, technology) %>%
+	  summarize_if(is.numeric , sum) %>%
```
Comment (bpbond) [inline R/zchunk_L104.bcoc_en_USA_S_T_Y.R]: Are you really summarizing multiple numeric columns here? If so, can they be listed explicitly? If not, something like `summarise(value = sum(value)` may be clearer.

File: R/zchunk_L104.bcoc_en_USA_S_T_Y.R [hunk 7]
```diff
@@ -32,49 +32,130 @@ module_emissions_L104.bcoc_en_USA_S_T_Y_DISABLED <- function(command, ...) {
...
+	  filter(!is.na(sector)) %>%
+	  rename(BCOC_agg_sector = sector) %>%
+	  ungroup() ->
+	  USA_bcoc_in_tg_1990
+
+	# Extract year 2000 Energy data
+	L101.in_EJ_R_en_Si_F_Yh %>%
+	  select(GCAM_region_ID, sector, technology, fuel, `2000`) %>%
+ 	  rename(Fuel_Use = `2000`) %>%
+ 	  filter(GCAM_region_ID == 1) %>%
```
Comment (bpbond) [inline R/zchunk_L104.bcoc_en_USA_S_T_Y.R]: There's a predefined constant for this in `constants.R` - something like `gcam.USA_REGION_ID`

File: R/zchunk_L104.bcoc_en_USA_S_T_Y.R [hunk 8]
```diff
@@ -32,49 +32,130 @@ module_emissions_L104.bcoc_en_USA_S_T_Y_DISABLED <- function(command, ...) {
...
+  	  # This is expected. Use left_join for this reason.
+	  left_join(Bond_bc.oc_tech,by = c("BCOC.Technology")) %>%
+	  group_by(sector, technology) %>%
+	  summarize_if(is.numeric , sum) %>%
+	  filter(!is.na(sector)) %>%
+	  rename(BCOC_agg_sector = sector) %>%
+	  ungroup() ->
+	  USA_bcoc_in_tg_1990
+
+	# Extract year 2000 Energy data
```
Comment (bpbond) [inline R/zchunk_L104.bcoc_en_USA_S_T_Y.R]: Why? What's special about 2000?

File: R/zchunk_L104.bcoc_en_USA_S_T_Y.R [hunk 9]
```diff
@@ -32,49 +32,130 @@ module_emissions_L104.bcoc_en_USA_S_T_Y_DISABLED <- function(command, ...) {
...
+
+    # Extract mapping between GCAM and Bond data
+ 	GCAM_sector_tech %>%
+ 	  select(sector, technology, fuel, BCOC_agg_sector, BCOC_agg_fuel) %>%
+ 	  filter(!is.na(BCOC_agg_sector)) -> # Where there is no correspondence remove those rows
+ 	  BCOC_GCAM_sector_tech
+
+    # Aggregate GCAM energy consumption to Bond BCOC categories
+ 	GCAM_USA_energy_2000 %>%
+      # Not all GCAM sectors have energy inputs, so use left_join since not all elements will have a match
```
Comment (bpbond) [inline R/zchunk_L104.bcoc_en_USA_S_T_Y.R]: Again great commenting

File: R/zchunk_L104.bcoc_en_USA_S_T_Y.R [hunk 10]
```diff
@@ -32,49 +32,130 @@ module_emissions_L104.bcoc_en_USA_S_T_Y_DISABLED <- function(command, ...) {
...
+      left_join(BCOC_GCAM_sector_tech,by=c("sector","technology","fuel")) %>%
```
Comment (bpbond) [inline R/zchunk_L104.bcoc_en_USA_S_T_Y.R]: Need spaces after `,` and around `=`

File: R/zchunk_L104.bcoc_en_USA_S_T_Y.R [hunk 11]
```diff
@@ -32,49 +32,130 @@ module_emissions_L104.bcoc_en_USA_S_T_Y_DISABLED <- function(command, ...) {
...
+  # Join using BCOC emissions first, since these are the categories we want data for.
+  # If there is no energy data, then want a zero emission factor for these categories
+  USA_bcoc_in_tg_1990 %>%
+    # Still don't expect a complete match, so use left_join
+    left_join(GCAM_energy_2000_byBCOC_agg,by=c("BCOC_agg_sector","technology")) %>%
+    mutate(bc_em_factor = BC_em / Fuel_Use) %>%
+    mutate(oc_em_factor = OC_em / Fuel_Use) %>%
+    select(-BC_em, -OC_em, -Fuel_Use) %>%
+    rename(sector = BCOC_agg_sector) %>%
+    mutate_all(funs(replace(., is.na(.), 0))) %>%
```
Comment (bpbond) [inline R/zchunk_L104.bcoc_en_USA_S_T_Y.R]: If possible please use `replace_na`

File: inst/extdata/emissions/Bond_BC_1990.csv [hunk 12]
```diff
@@ -0,0 +1 @@
...
UrbanBurn_OpenWaste,2.07,0.24,2.11,0.87,0.13,2.2,0.44,11.87,5.24,3.87,0.35,0.96,6.78,1.69,0.3,1.19,0.57
UrbanBurn_OpenBiomass,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
RuralBurn_OpenWaste,0.69,0.08,1.02,0,0,0.72,0.42,0,0,0,0,0,0,0,0,0,0
RuralBurn_OpenBiomass,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
SavannaBurn_OpenWaste,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
SavannaBurn_OpenBiomass,0,0,0.84,0,129.31,0.01,0,48.15,4.11,18.34,0,0.01,1121.92,124.83,4.33,262.8,0
ForestBurn_OpenWaste,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
ForestBurn_OpenBiomass,57.49,40.03,42.28,0.37,13.6,81.21,1.22,47.66,24.71,153.03,0.05,3.16,293.02,205.72,4.86,237.12,32.65
AgResidueBurn_OpenWaste,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
AgResidueBurn_OpenBiomass,3.02,12.25,10.72,1.28,20.7,16.32,4.32,78.49,57.72,34.03,1.19,2.93,50.62,11.98,5.58,14.05,2.37
```
Comment (bpbond) [inline inst/extdata/emissions/Bond_BC_1990.csv]: Here and below, please normalize line endings (or ping me and I'll do it). Thanks.

**Comment by ssmithClimate** [comment]
@bpbond for reviewer, Kate or Page.

File: inst/extdata/emissions/Bond_BC_1990.csv (hunk 12 above)
Comment (ssmithClimate) [inline inst/extdata/emissions/Bond_BC_1990.csv]: @bpbond Is this an issue of what standard are we using? Unix, Mac, DOS? 

Or are these in-consistent within the file?

File: R/zchunk_L104.bcoc_en_USA_S_T_Y.R (hunk 7 above)
Comment (ssmithClimate) [inline R/zchunk_L104.bcoc_en_USA_S_T_Y.R]: @bpbond Thanks, I had wondered about that.

File: inst/extdata/emissions/Bond_BC_1990.csv (hunk 12 above)
Comment (bpbond) [inline inst/extdata/emissions/Bond_BC_1990.csv]: We're just trying to standardize on [unix-y line endings](http://www.cs.toronto.edu/~krueger/csc209h/tut/line-endings.html).

File: R/zchunk_L104.bcoc_en_USA_S_T_Y.R (hunk 2 above)
Comment (ssmithClimate) [inline R/zchunk_L104.bcoc_en_USA_S_T_Y.R]: Good to know (although I'll probably leave this as is for clarity)

File: R/zchunk_L104.bcoc_en_USA_S_T_Y.R (hunk 6 above)
Comment (ssmithClimate) [inline R/zchunk_L104.bcoc_en_USA_S_T_Y.R]: Thanks, good idea. Done.

File: R/zchunk_L104.bcoc_en_USA_S_T_Y.R (hunk 8 above)
Comment (ssmithClimate) [inline R/zchunk_L104.bcoc_en_USA_S_T_Y.R]: Actually documented later in script, added note here pointing to that.

File: inst/extdata/emissions/Bond_BC_1990.csv (hunk 12 above)
Comment (kvcalvin) [inline inst/extdata/emissions/Bond_BC_1990.csv]: What are "Minicam 17 regions"? I don't remember MiniCAM having 17 regions.

Comment (ssmithClimate) [inline inst/extdata/emissions/Bond_BC_1990.csv]: It was before your time :)

Comment (ssmithClimate) [inline inst/extdata/emissions/Bond_BC_1990.csv]: File description changed to clarify

**Comment by bpbond** [comment]
👍 nice job @ssmithClimate - merging this

**Pull request merged**